// lastfm/index.js
/**
 * last.fm api model
 * @author adrienjoly
 */

var http = require('http');
var util = require('util');
//var crypto = require('crypto');

var apiKey = "9c5df33c08281a3dbd68378f4027728e";
var apiHost = "ws.audioscrobbler.com";
var apiPrefix = "/2.0/?api_key="+apiKey+"&format=json&method=";
/*
function md5(data) {
	return crypto.createHash('md5').update(data).digest("hex");
};

function hex(n, padding) {
	var str = n.toString(16);
	while (str.length < padding)
		str = '0' + str;
	return str;
}

function genAuth(path, obj) {
	var nc = hex(++ncount, 8);
	var a1 = [login, obj.realm, password].join(':');
	var a2 = "GET:"+path;
	var response = [md5(a1), obj.nonce, nc, cnonce, obj.qop, md5(a2)].join(':');
	var auth = [
		'Digest username="'+login+'"',
		'realm="'+obj.realm+'"',
		'nonce="'+obj.nonce+'"',
		'uri="'+path+'"',
		'algorithm='+obj.algorithm, // MD5
		'response="'+md5(response)+'"',
		'opaque="'+obj.opaque+'"',
		'qop=' + obj.qop, // auth
		'nc='+nc,
		'cnonce="'+cnonce+'"'
	].join(', ');
	console.log("-> generated auth", auth);
	return auth;
}

function parseRecvAuth(auth) {
	if (auth) {
		recvAuth = {};
		var defs = auth.substr(auth.indexOf(" ")+1).split(/,\s+/);
		for (var i in defs) {
			var def = defs[i].split('=');
			recvAuth[def[0]] = def[1].replace(/"/g, '');
		}
		console.log("-> received auth", recvAuth);
	}
}
*/
module.exports.submitRequest = function(suffix, cb) {
	util.log("submitting "+apiPrefix+suffix+" request to last.fm ...");
	return req = http.request({
		host: apiHost,
		path: apiPrefix + suffix,
		method: "GET",
	}, function (res) {
		res.setEncoding('utf-8');
		console.log("-> statusCode", res.statusCode);
		//if (res.statusCode == 401) // try again, with correct auth string
		//	!auth && exports.submitRequest(path, /*null*/"pouet", cb);
		//else
		{
			var json = "";
			res.addListener('data', function(chunk) {
				json += chunk.toString();
			});
			res.addListener('end', function() {
				try {
					json = JSON.parse(json);
				} catch(e) {};
				cb(json);
			});
		}
	})
	.end();
};

module.exports.getArtists = function(artistName, cb) {
	exports.submitRequest("artist.gettoptracks&artist="+artistName, function(res) {
		cb(res);
	});
}
